wagon-providers/wagon-http-shared/src/main/java/org/apache/maven/wagon/shared/http/AbstractHttpClientWagon.java

```diff
@@ -614,11 +800,23 @@ public void openConnectionInternal()
                                                                     getRepository().getPort() );
                 credentialsProvider.setCredentials( targetScope, creds );
             }
+            //MNG-5583 per endpoint PKI authentication
+            if ( authenticationInfo.getTrustStore() != null || authenticationInfo.getKeyStore() != null ) 
+            {
+                //cache the client for this specific server host:port combination
+                String key = repository.getProtocol() + repository.getHost() + repository.getPort();
+                if ( !httpClientWithCustomSslSocketFactoryCache.containsKey( key ) )
+                {
+                    httpClientWithCustomSslSocketFactoryCache.put( key, 
+                            initilaizeLocalHttpClientWithCustomSslSocketFactory() );
+                }
+            }
         }
 
         ProxyInfo proxyInfo = getProxyInfo( getRepository().getProtocol(), getRepository().getHost() );
         if ( proxyInfo != null )
         {
+            //TODO add PKI support?
```

Comment:
This would require two things:

* A proxy which speaks HTTPS as its protocol
* The proxy from above + TLS mutual auth supported

If this cannot be set up with OSS tools no need to add a TODO here. Maybe Apache's mod_proxy might be tried as forwarding proxy.

Reply:
i'm pretty sure it's possible, but i'll remove the todo if you like.